src/handler/sign_tx.c

```diff
@@ -1,102 +1,442 @@
 
-        return io_send_sw(SW_OK);
-    } else {  // parse transaction
-        if (G_context.req_type != CONFIRM_TRANSACTION) {
-            return io_send_sw(SW_BAD_STATE);
+        // the remainder of the first byte was used to represent the bip32 path length of the change path
+        if (buffer[0] > MAX_BIP32_PATH) {
+            THROW(SW_WRONG_DATA_LENGTH);
+        }
+
+        if (cdata->size - cdata->offset < 4*buffer[0]) {
+            THROW(SW_WRONG_DATA_LENGTH);
+        }
+        memmove(buffer+1, cdata->ptr + cdata->offset, 4 * buffer[0]);
+        buffer_seek_cur(cdata, 4*buffer[0]);
+
+        buffer_t bufdata = {
+            .ptr = buffer,
+            .size=1 + 4*MAX_BIP32_PATH,
+            .offset = 0
+        };
+
+        // buffer holds the serialized bip32 path that was read from cdata
+        if (!buffer_read_bip32_path(&bufdata, &G_context.tx_info.change_bip32_path)) {
+            THROW(SW_WRONG_DATA_LENGTH);
+        }
+    }
+}
+
+/**
+ * Read the info needed to parse the sighash_all data
+ *
+ * 2 bytes for tx version
+ * - 1 : HTR Transaction (Only supported version, currently)
+ *
+ * 1 byte for number of tokens
+ * 1 byte for number of inputs
+ * 1 byte for number of outputs
+ *
+ * Obs: This is only on the first call with chunk=0, so it should never fail for lack of data
+ *
+ * The output will be on the global context for sign tx (`tx_info`)
+ **/
+void read_tx_data(buffer_t *cdata) {
+    if (!(buffer_read_u16(cdata, &G_context.tx_info.tx_version, BE) && // read version bytes (Big Endian)
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_tokens) && // read number of tokens, inputs and outputs, respectively
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_inputs) &&
+        buffer_read_u8(cdata, &G_context.tx_info.outputs_len))) {
+        // if an error occurs reading
+        THROW(SW_WRONG_DATA_LENGTH);
+    }
+}
+
+/**
+ * Add all of the input (buffer_t) to the sighash_all buffer
+ * Does not move the input offset
+ **/
+void sighash_all_hash(buffer_t *cdata) {
+    // cx_hash returns the size of the hash after adding the data, we can safely ignore it
+
+    cx_hash(
+        &G_context.tx_info.sha256.header,   // hash context pointer
+        0,                                  // mode (supports: CX_LAST)
+        cdata->ptr + cdata->offset,         // Input data to add to current hash
+        cdata->size - cdata->offset,        // Length of input data
+        NULL, 0);                           // output (if flag CX_LAST was set)
+}
+
+/**
+ * Sign sighash_all with given bip32 path derived private key and send the signature
+ *
+ * If sighash_all is empty it means that we need to sha256d the given data to initialize it
+ **/
+bool sign_tx_with_key() {
+    // the bip32 path should already be initialized when calling this method.
+
+    if (G_context.bip32_path.length == 0 || G_context.bip32_path.path == (uint32_t*) NULL) {
+        return io_send_sw(SW_SIGNATURE_FAIL) > 0;
```

Comment:
It seems we should send the error and always return false. What do you think?

Reply:
This is expected since the `apdu_dispatcher` expects a return >=0 in case of success and a faillure for the `apdu_dispatcher` means something critical and the app has to exit (to prevent attacks or unhandled/unintended execution).